Add user last seen information

luismulinari opened the pull request:
## Description

This PR introduces a `Last Seen` user meta to the users, which is going to be used to report and block inactive users, improving security by reducing the surface attack.

## Inactive users blocked on login
<img width="40%" src="https://github.com/Automattic/vip-go-mu-plugins/assets/33168/a5cd16c4-057b-49e8-9dd0-92c11fd9db02">

## Inactive users on the Users page (network and site)
<img src="https://github.com/Automattic/vip-go-mu-plugins/assets/33168/178a4f37-4054-4748-8b04-e4229caf1bbc">

## Inactive users 
```
curl -X GET --user "example@example.org:key" "http://vip-testing-luismulinari-multi.vipdev.lndo.site/wp-json/wp/v2/users?context=edit"
{"code":"rest_forbidden_context","message":"Sorry, you are not allowed to list users.","data":{"status":401}}%
```


## Changelog Description
<!--
A description of the context of the change for a changelog. It should have a title, examples (if applicable), and why the change was made.

**Please keep the changelog title format same as in example below (### <Title>), as this is later used to generate the changelog entry title.**

Example for a plugin upgrade:

### Plugin Updated: Jetpack 9.2.1

We upgraded Jetpack 9.2 to Jetpack 9.2.1.

Not a lot of significant changes in this patch release, just bugfixes and compatibility improvements.
-->
## Pre-review checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change uses a rollout method to ease with deployment (if applicable - especially for large scale actions that require writes).
- [x] This change has relevant documentation additions / updates (if applicable).
- [x] I've created a changelog description that aligns with the provided examples.

## Pre-deploy checklist

- [ ] VIP staff: Ensure any alerts added/updated conform to internal standards (see internal documentation). 

## Steps to Test
1. Check out PR.
2. Add the following constants:
```
define( 'VIP_SECURITY_INACTIVE_USERS_ACTION', 'BLOCK' );
define( 'VIP_SECURITY_CONSIDER_USERS_INACTIVE_AFTER_DAYS', 30 );
```
3. Go to `wp-admin` > `Users`

luismulinari committed:
Add user last seen information

luismulinari committed:
Linting fixes

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,60 @@
+	$cache_key = LAST_SEEN_UPDATE_USER_META_CACHE_KEY_PREFIX . $current_user->ID;
+
+	try {
+		if ( wp_cache_get( $cache_key ) ) {
+			return;
+		}
+
+		update_user_meta( $current_user->ID, 'vip_last_seen', time() );
+
+		wp_cache_set( $cache_key, true, null, LAST_SEEN_UPDATE_USER_META_CACHE_TTL );
```
This would be less race-condition prone:

```
if ( wp_cache_add( $cache_key, true, 'vip', LAST_SEEN_UPDATE_USER_META_CACHE_TTL ) ) {
	update_user_meta( $current_user->ID, 'vip_last_seen', time() );
}
```

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,60 @@
+add_filter('manage_users_custom_column', 'users_custom_column', 10, 3 ) ;
+
+function update_user_last_seen() {
+	global $current_user;
+
+	$cache_key = LAST_SEEN_UPDATE_USER_META_CACHE_KEY_PREFIX . $current_user->ID;
```
Not 100% sure, but might be possible for the current user ID to be `0` if some code does `wp_set_current_user( 0 );`.  So might want to safeguard that and make sure we actually have an ID first, and that `$current_user` is set / not null / has that ID property.

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,60 @@
+
+		wp_cache_set( $cache_key, true, null, LAST_SEEN_UPDATE_USER_META_CACHE_TTL );
+	} catch ( \Exception $e ) {
+		trigger_error(
+			sprintf( 'failed to update user last seen meta', esc_html( $e->getMessage() ) ),
+			E_USER_WARNING
+		);
```
I like error handling, but unfortunately WP doesn't really throw when things fail like DB writes. So no need here for the try/catch, as there's nothing above that will throw.

luismulinari committed:
Code review suggestions

WPprodigy commented on security/user-last-seen.php:
Let's namespace this file to avoid global function name conflicts. Example: https://github.com/Automattic/vip-go-mu-plugins/blob/e1227255d3b0c48c62fbcce37e8159263d9f2d10/security/login-error.php

luismulinari commented:
Thanks for such a thoughtful review @WPprodigy  ❤️

luismulinari committed:
Lockdown inactive users

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Fix priority

luismulinari committed:
Fix filter

luismulinari committed:
Simplify code execution flow

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+	if ( $is_api_request && is_considered_inactive( $user_id ) ) {
+		// To block API requests for inactive requests, we need to return a WP_Error object here
+		return new \WP_Error( 'inactive_account', __( '<strong>Error</strong>: Your account has been flagged as inactive. Please contact your site administrator.', 'inactive-account-lockdown' ) );
+	}
```
Is there a reason we're not blocking non-API requests here?

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+	}
+
+	$unblock_link = '';
+	if ( current_user_can( 'edit_user', array() ) ) {
```
```suggestion
	if ( current_user_can( 'edit_user', $user_id ) ) {
```

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+
+add_action( 'user_row_actions', function ( $actions  ) {
+	if( isset($_GET['action'] ) && $_GET['action'] === 'reset_last_seen' ){
+		$user_id = $_GET['user_id'];
```
```suggestion
		$user_id = absint( $_GET['user_id'] );
```

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+add_action( 'user_row_actions', function ( $actions  ) {
+	if( isset($_GET['action'] ) && $_GET['action'] === 'reset_last_seen' ){
+		$user_id = $_GET['user_id'];
+		delete_user_meta( $user_id, LAST_SEEN_META_KEY );
```
This handling should probably be moved to an earlier hook like `admin_init`.

We also need to:

- validate the user exists
- check `current_user_can( 'edit_user', $user_id )`
- do a nonce check to prevent CORS

And then, after deleting, it would be nice to redirect back the users page with a success message.

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+		return $views;
+	}
+
+	$count = $wpdb->get_var( 'SELECT COUNT(meta_key) FROM ' . $wpdb->usermeta . ' WHERE meta_key = "' . LAST_SEEN_META_KEY . '" AND meta_value < ' . get_inactivity_timestamp() );
```
```suggestion
	$count = $wpdb->get_var( $wpdb->prepare( 'SELECT COUNT(meta_key) FROM ' . $wpdb->usermeta . ' WHERE meta_key = %s AND meta_value < %s', LAST_SEEN_META_KEY, get_inactivity_timestamp() ) );
```

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+	$view = __( 'Blocked Users' );
+	if ( $count ) {
+		$class = isset( $_REQUEST[ 'last_seen_filter' ] ) ? 'current' : '';
+		$view = '<a class="' . $class . '" href="users.php?last_seen_filter=blocked">' . $view . '</a>';
```
```suggestion
		$view = '<a class="' . esc_attr( $class ) . '" href="users.php?last_seen_filter=blocked">' . esc_html( $view ) . '</a>';
```

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+		return false;
+	}
+
+	$last_seen_timestamp = get_user_meta( $user_id, LAST_SEEN_META_KEY, true );
+	if ( ! $last_seen_timestamp ) {
+		return false;
+	}
```
If a user never logs in to their account, we should probably still deactivate them after the threshold.

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+const LAST_SEEN_META_KEY = 'wpvip_last_seen';
+const LAST_SEEN_UPDATE_USER_META_CACHE_TTL = MINUTE_IN_SECONDS * 5; // Store last seen once every five minute to avoid too many write DB operations
+const LAST_SEEN_UPDATE_USER_META_CACHE_KEY_PREFIX = 'wpvip_last_seen_update_user_meta_cache_key';
+const GROUP = 'wpvip';
```
An alternate way to do this would be set the group as `user_last_seen` and the key can just be the `$user_id`.

We may also want to change the group to a global one because one user accessing 5 different network sites at the same time would trigger 5 DB writes (instead of 1).

mjangda commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,148 @@
+		return $user_id;
+	}
+
+	$is_api_request = ( ( defined( 'XMLRPC_REQUEST' ) && XMLRPC_REQUEST ) || ( defined( 'REST_REQUEST' ) && REST_REQUEST ) );
```
If we add `use Automattic\VIP\Utils\Context` to the top, we can do this:

```suggestion
	$is_api_request = Context::is_rest_api() || Context::is_xmlrpc_api();
```

mjangda reviewed (changes requested):
Spotted a couple of things I wanted to flag.

luismulinari committed:
Apply suggestions from code review

Co-authored-by: Mohammad Jangda <mo@automattic.com>

luismulinari committed:
Apply suggestions from code review

Co-authored-by: Mohammad Jangda <mo@automattic.com>

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Add last seen to network users list

luismulinari committed:
Move the logic to a class

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Add unit tests

luismulinari committed:
Linting fixes

luismulinari committed:
Fix tests to run on WP 5.8

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Update test-user-last-seen.php

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,273 @@
+		add_action( 'admin_init', array( $this, 'register_release_date' ) );
+
+		add_filter( 'determine_current_user', array( $this, 'determine_current_user' ), 30, 1 );
+		add_filter( 'authenticate', array( $this, 'authenticate' ), 20, 1 );
```
Can this be brought down inside the conditional hooks below? `authenticate` looks like it can move down to being inside `$this->is_block_action_enabled()` rather than checking that condition inside the method itself.

luismulinari replied:
It makes total sense!

Implemented on https://github.com/Automattic/vip-go-mu-plugins/pull/4765/commits/bb49b0cb066c6e26aff2f395733629ecd457c1fb

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,273 @@
+			// Force current user to 0 to avoid recursive calls to this filter
+			wp_set_current_user( 0 );
+
+			return new \WP_Error( 'inactive_account', __( '<strong>Error</strong>: Your account has been flagged as inactive. Please contact your site administrator.', 'inactive-account-lockdown' ) );
```
For the translation text domain, I think maybe we should stick with `wpvip` to keep mu-plugins easier to translate all around. But we've not really kept that up super great.

If this intends to ever be a standalone plugin though, then keeping a specific text domain does make sense.

luismulinari replied:
Fixed all the translation domains to `wpvip` on https://github.com/Automattic/vip-go-mu-plugins/pull/4765/commits/bb49b0cb066c6e26aff2f395733629ecd457c1fb

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,273 @@
+				'count_total'  => true,
+			),
+		);
+		$count       = (int) $users_query->get_total();
```
This could potentially be very expensive, and is just for some UI text that isn't super required. Some thoughts:

1) Maybe we just query w/ a limit of 1 user, and if it finds one then we show the filters/buttons.
2) Query with a limit of 101, and show 100+ if it returns the full amount.
3) Drop the query altogether and just show a button without a count.

luismulinari replied:
Good idea!

I ended up choosing `1` since it seems to be the best cost/benefit for now.

Implemented on https://github.com/Automattic/vip-go-mu-plugins/pull/4765/commits/bb49b0cb066c6e26aff2f395733629ecd457c1fb

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,273 @@
+		if ( ! get_option( self::LAST_SEEN_RELEASE_DATE_TIMESTAMP_OPTION_KEY ) ) {
+			// Right after the first admin_init, set the release date timestamp
+			// to be used as a fallback for users that never logged in before.
+			add_option( self::LAST_SEEN_RELEASE_DATE_TIMESTAMP_OPTION_KEY, time() );
```
Let's make this a non-autoloaded option, since it's not something needed on every request.

luismulinari replied:
TIL about autoloaded options!

Implemented on https://github.com/Automattic/vip-go-mu-plugins/pull/4765/commits/bb49b0cb066c6e26aff2f395733629ecd457c1fb

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,273 @@
+	}
+
+	public function get_inactivity_timestamp() {
+		return strtotime( sprintf( '-%d days', constant( 'VIP_SECURITY_CONSIDER_USERS_INACTIVE_AFTER_DAYS' ) ) ) + self::LAST_SEEN_UPDATE_USER_META_CACHE_TTL;
```
Do we need to validate/sanitize `VIP_SECURITY_CONSIDER_USERS_INACTIVE_AFTER_DAYS` first here? Make sure it's a number string and if not fallback to some sane default?

luismulinari replied:
Good idea.

Implemented on https://github.com/Automattic/vip-go-mu-plugins/pull/4765/commits/bb49b0cb066c6e26aff2f395733629ecd457c1fb

WPprodigy reviewed (commented):
Left a few thoughts. And then one larger question is whether we want to be setting this meta on all user roles or maybe just the ones with elevated permissions?

For example, subscriber (core) and customer (woocommerce) don't seem like account types we should be blocking. Returning to an eCommerce site a year or two later and being blocked form ordering or checking receipts is not good.

It additionally would increase the metadata in the DB and make the user queries here more painful.

Rather than one-by-one allow-listing certain roles, maybe we look for specific caps like manage_options, publish_posts, edit_posts, etc.

luismulinari committed:
Code review suggestions

luismulinari committed:
Check user caps inside is_considered_inactive

luismulinari commented:
Hey @WPprodigy, thanks for this awesome review once again. It's been a joy to learn about WordPress from you :)

> one larger question is whether we want to be setting this meta on all user roles or maybe just the ones with elevated permissions?

It makes total sense. I implemented that as you suggested and also added a filter `vip_security_last_seen_elevated_capabilities` to be able to customize the elevated capabilities on the fly.

What do you think?

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

WPprodigy committed:
Couple tweaks

- Removed excess $this->is_block_action_enabled() call
- Don’t lookup the option for ajax requests (which can be logged-out FE users)
- VIP_SECURITY_CONSIDER_USERS_INACTIVE_AFTER_DAYS doesn’t have to be defined in order for the block action to be enabled. It defaults to 90 days.
- Re-organized the capabilities (most common first, more or less). Uses user_can() so that filters have a say in the matter.

WPprodigy commented on security/user-last-seen.php:
```diff
@@ -0,0 +1,330 @@
+				'fields'       => 'ID',
+				'meta_key'     => self::LAST_SEEN_META_KEY,
+				// phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+				'meta_value'   => $this->get_inactivity_timestamp(),
```
Because this number always changes, it's not able to take advantage of any query caching 😬 . TBD on performance for this on larger sites, but might want to re-visit layer and add a simple caching layer around it.

WPprodigy commented on security.php:
```diff
@@ -16,6 +16,13 @@
 require_once __DIR__ . '/security/login-error.php';
 require_once __DIR__ . '/security/password.php';
 
+if ( defined( 'VIP_SECURITY_INACTIVE_USERS_ACTION' ) ) {
+	require_once __DIR__ . '/security/user-last-seen.php';
+
+	$last_seen = new \Automattic\VIP\Security\User_Last_Seen();
+	$last_seen->init();
+}
```
I think maybe we should load all this on the `plugins_loaded` hook, to allow for sites to have more time to conditionally setup the constants and use the filters.

That said, need to make sure it's still early enough to catch any of the authenticate calls.

```
add_action( 'plugins_loaded', function() {
	if ( defined( 'VIP_SECURITY_INACTIVE_USERS_ACTION' ) ) {
		require_once __DIR__ . '/security/user-last-seen.php';

		$last_seen = new \Automattic\VIP\Security\User_Last_Seen();
		$last_seen->init();
	}
} );
```

WPprodigy reviewed (commented):
The latest changes look good, I altered the capabilities part a bit.

Additionally, when a new user is added or a user is promoted and we’re past the LAST_SEEN_RELEASE_DATE_TIMESTAMP_OPTION_KEY period, then they are gonna start out as being blocked I think?

luismulinari committed:
Introduce ignore inactivity check until

luismulinari committed:
Add extra time when user is promoted to prevent user lockdown

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

Commit:
Add tests; Fix linting

luismulinari committed:
Fix linting

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Avoid using determine_user filter

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

rebeccahum commented on security/class-user-last-seen.php:
```diff
@@ -0,0 +1,388 @@
+
+		add_action( 'admin_init', array( $this, 'register_release_date' ) );
+		add_action( 'set_user_role', array( $this, 'user_promoted' ) );
+		add_action( 'vip_support_user_added', function( $user_id ) {
```
Since we already have an existing job that clears out the `vip_support` users (https://github.com/Automattic/vip-go-mu-plugins/blob/66912f3ca4e270870250f35ab96d81c192b95a1a/vip-support/class-vip-support-user.php#L90) after a timeperiod of being added, do you think we can skip checking those types of users with that role for inactivity altogether?

rebeccahum commented on security/class-user-last-seen.php:
```diff
@@ -0,0 +1,388 @@
+			return;
+		}
+
+		$this->release_date = get_option( self::LAST_SEEN_RELEASE_DATE_TIMESTAMP_OPTION_KEY );
```
Are we using this property anywhere or just directly checking the option?

WPprodigy replied:
The options main purpose is for: https://github.com/Automattic/vip-go-mu-plugins/pull/4765/files#diff-d6cd56ee4fe98abf6b5d384fc94675502abd9ad6a5d412954a4ccccc35cb2bb4R311-R314

But I agree, let's just call get_option() when it's needed. There is runtime caching in place for it, and validation is handled internally already. So no need for the class property to maintain it's own state. As-is, the current version will run this DB query/memcache get on every request rather unnecessarily.

rebeccahum commented on security/class-user-last-seen.php:
```diff
@@ -0,0 +1,388 @@
+			$error = __( 'User not found.', 'wpvip' );
+		}
+
+		if ( ! current_user_can( 'edit_user', $user_id ) ) {
```
Minor, but can we move this permissions check up earlier?

luismulinari committed:
Register activity on set_current_user and rest_authentication_errors

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

WPprodigy committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Use the determine_current_user filter to record activity; Use the wp_is_application_passwords_available_for_user filter to block REST requests with application passwords

WPprodigy commented on security/class-user-last-seen.php:
```diff
@@ -94,23 +99,35 @@ public function authenticate( $user ) {
+	 * @return bool
+	 */
+	public function application_password_authentication( $available, $user ) {
+		global $wp_last_seen_application_password_error;
```
Let's use a class property rather than a global

luismulinari replied:
Good call. Done: 182dc0d33

luismulinari committed:
Use a class property rather than a global

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

nickdaugherty commented:
I have a higher level thought on this ... can we reduce the instances of VIP-specific naming inside the class (meta/option keys, filter names) and make it more generic? This functionality is something that has wider appeal outside of VIP and by removing VIP references, re create an easier path to open sourcing this as a standalone plugin one day. It's always a goal to contribute back when possible and this seems like it can be a great standalone plugin in the future.

nickdaugherty commented on security/class-user-last-seen.php:
```diff
@@ -0,0 +1,420 @@
+		 *
+		 * @param array $skip_users The list of user IDs to skip.
+		 */
+		$skip_users = apply_filters( 'vip_security_last_seen_skip_users', array() );
```
What do you think about moving this filter to `is_considered_inactive()`? That gives more flexibility for filtering the results of the check and might be more predictable.

nickdaugherty commented on security/class-user-last-seen.php:
```diff
@@ -0,0 +1,420 @@
+
+	public function ignore_inactivity_check_for_user( $user_id, $until_timestamp = null ) {
+		if ( ! $until_timestamp ) {
+			$until_timestamp = strtotime( '+2 days' );
```
The default of `+2 days` feels a little too magic and is hard to predict. Maybe we should just require `$until_timestamp` be specified?

nickdaugherty replied:
Minor, but I'd also consider naming the function `ignore_inactivity_check_for_user_until( $user_id, $until_timestamp )`. The `_until` makes it obvious that it's a limited time action, vs. permanent (which is what I thought it would do before I read the code)

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

rinatkhaziev committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Check if constant is defined before using it

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

luismulinari committed:
Merge branch 'develop' into add/users-last-seen

rinatkhaziev committed:
Merge branch 'develop' into add/users-last-seen
